dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/workflow/artifacts/WorkflowArtifactServiceBean.java

```diff
@@ -0,0 +1,151 @@
+    private Map<StorageType, StorageService> services = new HashMap<>();
+
+    private WorkflowArtifactRepository repository;
+
+    // -------------------- CONSTRUCTORS --------------------
+
+    public WorkflowArtifactServiceBean() { }
+
+    @Inject
+    public WorkflowArtifactServiceBean(WorkflowArtifactRepository repository) {
+        this.repository = repository;
+    }
+
+    // -------------------- LOGIC --------------------
+
+    /**
+     * Registers service based on service's {@link StorageType}. If there is already a registered service supporting
+     * this type of storage, it would be unregistered.
+     * @param service service to be registered
+     */
+    public void register(StorageService service) {
+        StorageType storageType = Objects.requireNonNull(service.getStorageType());
+        services.put(storageType, service);
+    }
+
+    /**
+     * Same as {@link WorkflowArtifactServiceBean#saveArtifact(DatasetVersion, WorkflowExecutionStep, ArtifactData, StorageType)}
+     * but with the last parameter set to {@link StorageType#DATABASE}.
+     */
+    @TransactionAttribute(TransactionAttributeType.REQUIRES_NEW)
+    public WorkflowArtifact saveArtifact(DatasetVersion version, WorkflowExecutionStep executionStep, ArtifactData data) {
+        return saveArtifact(version, executionStep, data, StorageType.DATABASE);
+    }
+
+    /**
+     * Saves artifact into storage of selected type. The {@link WorkflowExecutionStep} parameter is optional (ie. could
+     * be null), but recommended to set.
+     * Please note that in case of text streams proper encoding is up to caller.
+     */
+    @TransactionAttribute(TransactionAttributeType.REQUIRES_NEW)
+    public WorkflowArtifact saveArtifact(DatasetVersion version, WorkflowExecutionStep executionStep, ArtifactData data, StorageType storageType) {
+        StorageService service = services.get(storageType);
+        WorkflowArtifact workflowArtifact = new WorkflowArtifact();
+        workflowArtifact.setDatasetVersionId(version.getId());
+        Optional.ofNullable(executionStep)
+                .map(WorkflowExecutionStep::getId)
+                .ifPresent(workflowArtifact::setWorkflowExecutionStepId);
+        workflowArtifact.setEncoding(data.getEncoding());
+        workflowArtifact.setArtifactName(data.getName());
+        workflowArtifact.setStorageType(storageType.name());
+        workflowArtifact = repository.save(workflowArtifact);
+
+        String location = service.save(workflowArtifact, data.inputStreamSupplier);
+
+        workflowArtifact.setLocation(location);
+        workflowArtifact.setCreatedAt(Instant.now());
+        return repository.save(workflowArtifact);
+    }
+
+    /**
+     * Returns {@link Optional} containing {@link InputStream} of stored data for the given
+     * {@link WorkflowArtifact} or empty {@link Optional} if value was not found.
+     */
+    public Optional<InputStream> readAsStream(WorkflowArtifact artifact) {
```

Comment:
I don't think we need the whole object of `WorkflowArtifact` to get it's binary data here. When I fetch the list of artifacts for given workflow execution I get a list of objects containing all the data required to go directly into the storage - they have the storage type and the storage location. Hence I can render links that  would contain both of them and decouple storage handling from artifact handling entirely. I think it'd be enough for this method to accept just a storage type and a location parameters instead.